<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print" />

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Mapdamage by ginolhac</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Mapdamage</h1>
        <h2>mapDamage: tracking and quantifying damage patterns in ancient DNA sequences</h2>
        <a href="https://github.com/ginolhac/mapDamage" class="button"><small>View project on</small>GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h2>Introduction</h2>

<p>mapDamage2 is a computational framework written in Python and R, which tracks and quantifies DNA damage patterns 
among ancient DNA sequencing reads generated by Next-Generation Sequencing platforms. </p>

<h3>Requirements</h3>

<ul>
<li>Python (version &gt;= 2.6)</li>
<li>
<a href="http://git-scm.com/">Git</a> </li>
<li>
<a>R</a> (version &gt;= 2.15.1) must be present in your $PATH. Otherwise, only tables will be produced.</li>
<li>
<a href="http://code.google.com/p/pysam/%5D">pysam</a> (version == 0.6) python package, an interface for reading/writing SAM/BAM files</li>
<li>
<a href="http://www.gnu.org/software/gsl/">gsl</a> GNU scientific library (GSL) </li>
<li>R libraries:

<ul>
<li>inline</li>
<li>gam</li>
<li>Rccp</li>
<li>RcppGSL</li>
<li>ggplot2 (&gt;=0.9.2)</li>
</ul>
</li>
</ul><p>Of note, the R libraries and the GSL library are only required for the statistical estimation of DNA damage 
parameters. seqtk is included in this package for fast computation of reference base composition, it is written by Heng Li
and can be found at <a href="https://github.com/lh3/seqtk">here at github.</a></p>

<h3>Installation</h3>

<p>mapDamage2 was successfully tested on GNU/Linux and MacOSX environments.</p>

<ol>
<li><p>Install mapDamage:<br>
Clone the repository from GitHub<br><code>git clone https://github.com/ginolhac/mapDamage.git mapDamage</code><br>
go into the resulting folder<br><code>cd mapDamage</code><br>
then clone the submodules.<br><code>git submodule update --init</code><br>
Run the following command if you have administrator rights<br><code>sudo python setup.py install</code><br>
otherwise install it locally<br><code>python setup.py install --user</code><br>
then, $HOME/.local/bin must be in your PATH.</p></li>
<li><p>Install pysam 0.6:<br>
Download the tarball archive for <a href="http://code.google.com/p/pysam/downloads/list">pysam</a>.
Untar the downloaded tar file and follow instructions in the pysam-0.6/INSTALL text file.
<strong>It is important installing the specific version pysam 0.6 since we had issues with newer versions.</strong>
It is recommend to check if the pysam installation seems okay before proceeding with the following command<br><code>python -c "import pysam"</code><br>
If nothing appears then proceed.</p></li>
<li><p>Install R:<br>
Follow the instructions available at <a href="http://www.r-project.org/">R project</a>
(with installers/packages available for all major operating systems).</p></li>
<li><p>Install the R packages:<br>
By running in a R console and selecting a CRAN mirror:<br><code>install.packages("inline")</code><br><code>install.packages("gam")</code><br><code>install.packages("Rcpp")</code><br><code>install.packages("ggplot2")</code> </p></li>
<li><p>Install the  GSL library <a href="http://www.gnu.org/software/gsl/">gsl</a>:<br>
For Debian-based distros by the following.<br><code>sudo apt-get install libgsl0-dev</code><br>
There should be a equivalent package for Mac using fink or macports.</p></li>
<li><p>Install RcppGSL:<br>
In a R session do the following<br><code>install.packages("RcppGSL")</code><br>
If the installation of RcppGSL ended with an unsuccessfully then, download the tarball at <a href="http://cran.r-project.org/web/packages/RcppGSL/index.html">RcppGSL</a> and install it with:<br><code>install.packages("RcppGSL_0.2.0.tar.gz")</code></p></li>
</ol><p>Note that if steps 4-6 fail for some reason, then it is possible to utilize mapDamage without 
the statistical function.</p>

<hr><h2>News in version 2.0</h2>

<ul>
<li>The main new feature is the <strong>approximate bayesian estimation of damage parameters</strong>.</li>
<li>The --rescale parameter can be optionally used to rescale quality scores of likely damaged positions 
in the reads. A new BAM file is constructed by downscaling quality values for misincorporations likely 
due to ancient DNA damage according to their initial qualities, position in reads and damage patterns.</li>
<li>The program was rewritten in python to make it simpler and the dependency on bedtools was replaced by the use of pysam.
The memory footprint is now negligible and runtime reduced by <em>ca.</em> 25%.</li>
<li>It is now possible to filter out nucleotides with qualities inferior to a threshold defined by users, 
allowing users to reduce the impact of sequencing errors when counting misincorporations.</li>
</ul><h2>Inputs</h2>

<p>Two files are needed:</p>

<ul>
<li>A valid SAM/BAM file with a correct header, as argument to the <strong>-i</strong> option.</li>
<li>A FASTA file that contains reference sequences used for mapping reads, as argument to the <strong>-r</strong> option. </li>
</ul><p>References described in the SAM/BAM header and the FASTA file must be coherent, <em>i.e</em>,<br>
the references must have identical names and lengths. Extra sequences present in the FASTA header<br>
raise a warning but the program will proceed since all needed references are available.</p>

<p>As an alternative, one can run only the plotting, statistic estimations or rescaling
on an already processed dataset. Use a combination of <strong>-d</strong> option followed by 
a valid folder and the <code>--plot-only</code>, <code>--stats-only</code> or  <code>--rescale-only</code> options.</p>

<h2>Outputs</h2>

<p>When all options are activated, 15 files are produced in the result folder.</p>

<ul>
<li>Runtime_log.txt, log file with a summary of command lines used and timestamps.</li>
</ul><p>For the plotting:</p>

<ul>
<li>a pdf file, Fragmisincorporation_plot.pdf that displays both fragmentation and misincorporation patterns.</li>
<li>a pdf file, Length_plot.pdf, that displays length distribution of singleton reads per strand and cumulative 
frequencies of C&gt;T at 5'-end and G&gt;A at 3'-end are also displayed per strand.</li>
<li>misincorporation.txt, contains a table with occurrences for each type of mutations and relative positions from the reads ends.</li>
<li>5pCtoT_freq.txt, contains frequencies of Cytosine to Thymine mutations per position from the 5'-ends.</li>
<li>3pGtoA_freq.txt, contains frequencies of Guanine to Adenine mutations per position from the 3'-ends.</li>
<li>dnacomp.txt, contains a table of the reference genome base composition.</li>
<li>lgdistribution.txt, contains a table with read length distributions per strand.</li>
</ul><p>For the statistical estimation:</p>

<ul>
<li>Stats_out_MCMC_hist.pdf, MCMC histogram for the damage parameters and log likelihood.</li>
<li>Stats_out_MCMC_iter.csv, values for the damage parameters and log likelihood in each MCMC iteration.</li>
<li>Stats_out_MCMC_trace.pdf, a MCMC trace plot for the damage parameters and log likelihood. </li>
<li>Stats_out_MCMC_iter_summ_stat.csv, summary statistics for the damage parameters MCMC samples.</li>
<li>Stats_out_post_pred.pdf, empirical misincorporation frequency and predictive intervals from the fitted model.</li>
<li>Stats_out_MCMC_correct_prob.csv, position specific probability of a C.T and G.A misincorporation is due to damage.</li>
<li>Rescaled BAM file, where likely post-mortem damaged bases have downscaled quality scores. </li>
</ul><h2>Examples and datasets</h2>

<p>A simple command line that would process the entire BAM file with plotting and statistic estimation is:</p>

<pre><code>mapDamage -i mymap.bam -r myreference.fasta
</code></pre>

<p>To run the plotting part with a new scale to fit lower levels of DNA damages, 0.1 as y-limit instead of 0.3 by default:</p>

<pre><code> mapDamage -d results_mydata -y 0.1 --plot-only
</code></pre>

<p>To run the statistic estimations using only the 5'-ends and with verbose output:</p>

<pre><code> mapDamage -d results_mydata --forward --stats-only -v
</code></pre>

<p>The original web page with examples, datasets and result files is there:</p>

<p><a href="http://geogenetics.ku.dk/publications/mapdamage/">http://geogenetics.ku.dk/publications/mapdamage/</a></p>

<h2>Usage</h2>

<pre><code>Usage: mapDamage [options] -i BAMfile -r reference.fasta

Use option -h or --help for help

Options:

   --version             show program's version number and exit
   -h, --help            show this help message and exit

Input files:

    -i FILENAME, --input=FILENAME
                         SAM/BAM file, must contain a valid header, use '-' for reading a BAM from stdin
    -r REF, --reference=REF
                        Reference file in FASTA format
General options:
    -n DOWNSAMPLE, --downsample=DOWNSAMPLE
                        Downsample to a randomly selected fraction of the reads (if 0 &lt; DOWNSAMPLE &lt; 1), 
                        or a fixed number of randomly selected reads (if DOWNSAMPLE &gt;= 1). By default, 
                        no downsampling is performed.
--downsample-seed   Seed value to use for downsampling. See documentation for py module 'random' for default behavior.
    -l LENGTH, --length=LENGTH
                        read length, in nucleotides to consider [70]
    -a AROUND, --around=AROUND
                        nucleotides to retrieve before/after reads [10]
    -Q MINQUAL, --min-basequal=MINQUAL
                        minimun base quality Phred score considered, Phred-33 assumed [0]
    -d FOLDER, --folder=FOLDER
                    folder name to store results [results_FILENAME]
    -f, --fasta         Write alignments in a FASTA file
    --plot-only         Run only plotting from a valid result folder
    -q, --quiet         Disable any output to stdout
    -v, --verbose       Display progression information during parsing

Options for graphics:

    -y YMAX, --ymax=YMAX
                        graphical y-axis limit for nucleotide misincorporation frequencies [0.3]
    -m READPLOT, --readplot=READPLOT
                        read length, in nucleotides, considered for plotting nucleotide 
                        misincorporations [25]
    -b REFPLOT, --refplot=REFPLOT
                        the number of reference nucleotides to consider for ploting base 
                        composition in the region located upstream and downstream of 
                        every read [10]
    -t TITLE, --title=TITLE
                        title used for both graph and filename [plot]

Options for the statistical estimation:

    --rand=RAND         Number of random starting points for the likelihood optimization [30]
    --burn=BURN         Number of burnin iterations [10000]
    --adjust=ADJUST     Number of adjust proposal variance parameters iterations [10]
    --iter=ITER         Number of final MCMC iterations [50000]
    --forward=FORWARD   Using only the 5' end of the seqs [False]
    --reverse=REVERSE   Using only the 3' end of the seqs [False]
    --fix-disp=FIX_DISP Fix dispersion in the overhangs [True]
    --diff-hangs        The overhangs are different for 5' and 3'  [False]
    --fix-nicks=FIX_NICKS 
                        Fix the nick frequency vector nu else estimate it with GAM [False]
    --jukes-cantor      Use Jukes Cantor instead of HKY85
    --single_stranded=SINGLE_STRANDED
                        Single stranded protocol [False]
    --seq-length=SEQ_LENGTH
                        How long sequence to use from each side [12]
    --stats-only        Run only statistical estimation from a valid result folder
    --rescale           Rescale the quality scores in the BAM file using the output from
                        the statistical estimation
    --rescale-only      Run only rescaling from a valid result folder
    --no-stats          Disabled statistical estimation, active by default
</code></pre>

<h2>Description of tables</h2>

<h3>misincorporation table</h3>

<p>This file looks like:</p>

<pre><code># table produced by mapDamage version 2
# using mapped file hits_sort_mts.bam and NC_012920.fasta as reference file
# Chr: reference from sam/bam header, End: from which termini of DNA sequences, Std: strand of reads
Chr End Std Pos A   C   G   T   Total   G&gt;A C&gt;T A&gt;G T&gt;C A&gt;C A&gt;T C&gt;G C&gt;A T&gt;G T&gt;A G&gt;C G&gt;T A&gt;- T&gt;- C&gt;- G&gt;- -&gt;A -&gt;T -&gt;C -&gt;G S
NC_012920.1 3p  +   1   424 579 201 424 1628    48  16  3   5   1   0   2   1   0   2   1   0   0   0   0   0   4   4   1   1   0
NC_012920.1 3p  +   2   466 535 213 404 1618    30  12  2   1   0   1   0   0   3   3   1   0   0   0   0   0   2   6   5   7   0
NC_012920.1 3p  +   3   514 463 223 417 1617    35  15  3   1   1   0   0   0   1   0   1   1   0   0   0   2   9   4   4   4   0
NC_012920.1 3p  +   4   514 483 221 400 1618    30  11  3   4   0   1   1   0   0   1   0   1   1   0   0   2   5   5   8   2   0
</code></pre>

<p>The first lines that start by a hash contain information about the options used while processing the data.
Then, the table contains occurrences for each of the 12 mutation type + 4 deletions + 4 insertions + 
soft-clipping, per reference (<code>Chr</code> column), per strand (<code>Std</code> column) , per end (<code>End</code> column) and per position (<code>Pos</code> column). 
In Z&gt;X, Z comes from the reference, X is the nucleotide from reads. Eventually, A, C, G, T come from the reference
and Total is the sum of the four nucleotides. Frequencies depicted in Fragmisincorporation_plot.pdf file are computed per
position as follows</p>

<p>occurrences of mutations / occurrences of the reference nucleotide </p>

<p>in order to compensate for base composition bias.</p>

<p>In this example, for G&gt;A at the first base from the 3'-ends and for the positive strand:
<code>48 / 201 = 0.238806</code></p>

<p>The Fragmisincorporation plot sums up misincorporations  for all references and strand orientations, finally 
displays the frequencies per position.  </p>

<p><strong>5p</strong> means reads analyzed from their 5'-ends and are depicted at the bottom left.</p>

<p><strong>3p</strong> means reads analyzed from their 3'-ends and are depicted at the bottom right.</p>

<p><strong>Remark concerning soft-clipping</strong>:</p>

<p>Soft-clipped bases are bases located at read extremities that are NOT aligned to the reference. 
Those bases are not taken into account and alignments are computed on aligned bases. 
However, the soft-clipped bases are recorded and displayed as an orange line as a diagnostic for users.
Even if, the positions for those bases are different from positions of all misincorporations, they should warn
users that reads should be trimmed if soft-clipped base frequencies is high.</p>

<h3>dnacomp table</h3>

<p>Example table below:</p>

<pre><code># table produced by mapDamage version 0.4.0
# using mapped file hits_sort_mts.bam and NC_012920.fasta as reference file
# Chr: reference from sam/bam header, End: from which termini of DNA sequences, Std: strand of reads
Chr End Std Pos A   C   G   T   Total
NC_012920.1 3p  +   -70 113 77  43  181 414
NC_012920.1 3p  +   -69 178 140 97  233 648
NC_012920.1 3p  +   -68 230 145 98  219 692
NC_012920.1 3p  +   -67 217 157 76  272 722
NC_012920.1 3p  +   -66 224 188 88  239 739
NC_012920.1 3p  +   -65 244 165 104 262 775
NC_012920.1 3p  +   -64 223 185 102 288 798
</code></pre>

<p>The first lines that start by a hash contain information about the options used while processing the data.
Then, the table contains occurrences for each nucleotides, per reference (<code>Chr</code> column), 
per strand (<code>Std</code> column) , per end (<code>End</code> column) and per position (<code>Pos</code> column).</p>

<p>At a given position, base composition is recorded either for the reference or for reads depending 
on the position and direction. The following table details which sequences and coordinates are 
used by default:</p>

<table>
<tr>
<th align="center">Positions</th>
<th align="center">5p</th>
<th align="center">3p</th>
</tr>
<tr>
<td align="center">negative values</td>
<td align="center">reference, -10 to -1</td>
<td align="center">reads, -25 to -1</td>
</tr>
<tr>
<td align="center">positive values</td>
<td align="center">reads, +1 to +25</td>
<td align="center">reference, +1 to +10</td>
</tr>
</table><h2>Citation</h2>

<p>If you use this program, please cite the following publication:<br>
Jónsson H, Ginolhac A, Schubert M, Johnson P, Orlando L.
mapDamage2.0: fast approximate Bayesian estimates of ancient DNA damage parameters.
<em>Bioinformatics</em> 2013. <em>in press</em></p>

<p>The original mapDamage1 is described in the following article:<br>
Ginolhac A, Rasmussen M, Gilbert MT, Willerslev E, Orlando L.
mapDamage: testing for damage patterns in ancient DNA sequences. <em>Bioinformatics</em> 2011 <strong>27</strong>(15):2153-5
<a href="http://bioinformatics.oxfordjournals.org/content/27/15/2153">http://bioinformatics.oxfordjournals.org/content/27/15/2153</a></p>

<h2>FAQ</h2>

<p>Q: I got this error: could not find function "ggtitle"</p>

<p>A: update your ggplot2 package to at least version 0.9.2 by running
update.packages("ggplot2") and selecting a CRAN mirror</p>

<p>Q: How to clone the development version from github?</p>

<p>A: Use the following commands</p>

<blockquote>
<p><code>git clone https://github.com/ginolhac/mapDamage.git mapDamage</code><br><code>cd mapDamage</code> 
<code>git submodule update --init</code>  </p>
</blockquote>

<h2>Contact</h2>

<p>Please report bugs and suggest possible improvements to Aurélien Ginolhac, Mikkel Schubert or Hákon Jónsson by email:
aginolhac at snm.ku.dk, MSchubert at snm.ku.dk or jonsson.hakon at gmail.com.</p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/ginolhac/mapDamage/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/ginolhac/mapDamage/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/ginolhac/mapDamage"></a> is maintained by <a href="https://github.com/ginolhac">ginolhac</a>.</p>

          <p>This page was generated by <a href="pages.github.com">GitHub Pages</a> using the Architect theme by <a href="http://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>